File: php/laravel/src/Transport/AsyncHttpTransport.php
<?php

namespace Smartness\TraceFlow\Transport;

use GuzzleHttp\Client;
use GuzzleHttp\Exception\GuzzleException;
use GuzzleHttp\Promise\Utils;

class AsyncHttpTransport extends AbstractHttpTransport
{
    /** @var \GuzzleHttp\Promise\PromiseInterface[] */
    private array $promises = [];

    /**
     * Last pending promise per entity order key. Used to serialize requests for
     * the same entity (e.g. a step's create then update) so an update can never
     * be transferred before its create completes, which would 404 on the server.
     *
     * @var array<string, \GuzzleHttp\Promise\PromiseInterface>
     */
    private array $chains = [];

    protected function buildClient(array $options): Client
    {
        try {
            return new Client($options);
        } catch (\Exception $e) {
            if ($this->silentErrors) {
                error_log("[TraceFlow Async] Error initializing client (silenced): {$e->getMessage()}");
                unset($options['base_uri']);

                return new Client($options);
            }
            throw $e;
        }
    }

    protected function dispatch(string $method, string $uri, array $payload, ?string $orderKey = null): void
    {
        $this->executeAsync($method, $uri, $payload, $orderKey);
    }

    protected function logPrefix(): string
    {
        return '[TraceFlow Async]';
    }

    private function executeAsync(string $method, string $uri, array $data, ?string $orderKey = null, int $attempt = 0): void
    {
        // The actual request is wrapped in a closure so it can be deferred until
        // any earlier request for the same entity has settled (see $chains).
        $send = function () use ($method, $uri, $data, $orderKey, $attempt) {
            return $this->client->requestAsync($method, $uri, ['json' => $data])
                ->then(
                    function ($response) {
                        // Request succeeded
                    },
                    function (GuzzleException $exception) use ($method, $uri, $data, $orderKey, $attempt) {
                        if ($attempt < $this->maxRetries) {
                            // Intentionally no delay between async retries: sleeping inside a promise
                            // rejection handler would block the event loop. The flush() while-loop
                            // naturally adds a small gap between retry batches. retry_delay config
                            // applies only to the synchronous HttpTransport.
                            $this->executeAsync($method, $uri, $data, $orderKey, $attempt + 1);
                        } else {
                            $this->recordFailure();
                            if ($this->silentErrors) {
                                error_log($this->logPrefix()." Failed after {$this->maxRetries} retries: {$exception->getMessage()}");
                            } else {
                                throw $exception;
                            }
                        }
                    }
                );
        };

        if ($orderKey !== null && isset($this->chains[$orderKey])) {
            // Serialize behind the previous request for this entity. Run regardless
            // of whether that request fulfilled or rejected so a failed create does
            // not permanently stall the entity's later requests.
            $promise = $this->chains[$orderKey]->then($send, $send);
        } else {
            $promise = $send();
        }

        if ($orderKey !== null) {
            $this->chains[$orderKey] = $promise;
        }

        $this->promises[] = $promise;
    }

    /**
     * Flush all pending async requests, including any retry promises created during settlement.
     */
    public function flush(): void
    {
        while (! empty($this->promises)) {
            $batch = $this->promises;
            $this->promises = [];

            try {
                Utils::settle($batch)->wait();
            } catch (\Exception $e) {
                if ($this->silentErrors) {
                    error_log($this->logPrefix()." Error during flush (silenced): {$e->getMessage()}");
                } else {
                    throw $e;
                }
            }
        }

        // All promises have settled; drop references to settled per-entity chains
        // so the map does not grow for the lifetime of the process.
        $this->chains = [];

        $this->drainQueue();
    }

    public function shutdown(): void
    {
        $this->flush();
    }
}

File: java/core/src/main/java/com/smartness/traceflow/transport/AsyncHttpTransport.java
package com.smartness.traceflow.transport;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.smartness.traceflow.TraceFlowConfig;
import com.smartness.traceflow.dto.TraceEvent;
import com.smartness.traceflow.exception.TraceFlowException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.Map;
import java.util.Queue;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentLinkedQueue;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.Supplier;

public final class AsyncHttpTransport implements Transport {

    private static final Logger log = LoggerFactory.getLogger(AsyncHttpTransport.class);

    private final HttpClient httpClient;
    private final String endpoint;
    private final String apiKey;
    private final boolean silentErrors;
    private final ObjectMapper objectMapper;
    private final RetryExecutor retryExecutor;
    private final Queue<CompletableFuture<?>> futures = new ConcurrentLinkedQueue<>();
    private final AtomicInteger eventCount = new AtomicInteger(0);

    /**
     * Last pending request per entity order key. Requests for the same entity
     * (e.g. a step's create then update) are chained so an update is never sent
     * before its create completes, which would 404 on the server. Independent
     * entities — and logs, which use a null key — still run concurrently.
     */
    private final Map<String, CompletableFuture<?>> chains = new ConcurrentHashMap<>();

    public AsyncHttpTransport(TraceFlowConfig config) {
        this.endpoint = config.endpoint();
        this.apiKey = config.apiKey();
        this.silentErrors = config.silentErrors();
        this.objectMapper = new ObjectMapper();
        this.retryExecutor = new RetryExecutor(config.maxRetries(), config.retryDelayMs());
        this.httpClient = HttpClient.newBuilder()
                .connectTimeout(config.timeout())
                .build();
    }

    @Override
    public void send(TraceEvent event) {
        try {
            EventRouter.Route route = EventRouter.route(event);
            String orderKey = orderKey(event);

            // The request (with retries and silent-error handling) is wrapped in a
            // supplier so it can be deferred until any earlier request for the same
            // entity has completed.
            Supplier<CompletableFuture<?>> request = () -> retryExecutor.executeAsync(() -> executeRequest(route))
                    .thenRun(() -> eventCount.incrementAndGet())
                    .exceptionally(ex -> {
                        if (silentErrors) {
                            log.warn("[TraceFlow Async] Error sending event (silenced): {}", ex.getMessage());
                        } else {
                            throw new TraceFlowException("Failed to send event", ex);
                        }
                        return null;
                    });

            final CompletableFuture<?> future;
            if (orderKey == null) {
                future = request.get();
            } else {
                // Atomically chain behind the previous request for this entity.
                // handle(...) detaches so the next request still runs even if the
                // previous one failed; the previous future stays in `futures` so its
                // outcome is still observed at flush time.
                future = chains.compute(orderKey, (key, previous) ->
                        previous == null
                                ? request.get()
                                : previous.handle((r, ex) -> null).thenCompose(ignored -> request.get()));
            }

            futures.add(future);
            future.whenComplete((r, e) -> futures.remove(future));
        } catch (Exception e) {
            if (silentErrors) {
                log.warn("[TraceFlow Async] Error sending event (silenced): {}", e.getMessage());
            } else {
                throw new TraceFlowException("Failed to send event", e);
            }
        }
    }

    /**
     * Order key identifying the entity a request belongs to, or null for events
     * that need no ordering (logs).
     */
    private static String orderKey(TraceEvent event) {
        return switch (event.eventType()) {
            case TRACE_STARTED, TRACE_FINISHED, TRACE_FAILED, TRACE_CANCELLED ->
                    "trace:" + event.traceId();
            case STEP_STARTED, STEP_FINISHED, STEP_FAILED ->
                    "step:" + event.traceId() + ":" + event.stepId();
            case LOG_EMITTED -> null;
        };
    }

    private CompletableFuture<Void> executeRequest(EventRouter.Route route) {
        String json;
        try {
            json = objectMapper.writeValueAsString(route.payload());
        } catch (Exception e) {
            return CompletableFuture.failedFuture(new TraceFlowException("Serialization failed", e));
        }

        HttpRequest.Builder requestBuilder = HttpRequest.newBuilder()
                .uri(URI.create(endpoint + route.path()))
                .header("Content-Type", "application/json");

        if (apiKey != null) {
            requestBuilder.header("X-API-Key", apiKey);
        }

        HttpRequest request = switch (route.method()) {
            case "POST" -> requestBuilder.POST(HttpRequest.BodyPublishers.ofString(json)).build();
            case "PATCH" -> requestBuilder.method("PATCH", HttpRequest.BodyPublishers.ofString(json)).build();
            default -> throw new TraceFlowException("Unsupported HTTP method: " + route.method());
        };

        return httpClient.sendAsync(request, HttpResponse.BodyHandlers.ofString())
                .thenAccept(response -> {
                    if (response.statusCode() >= 400) {
                        throw new TraceFlowException("HTTP " + response.statusCode() + ": " + response.body());
                    }
                });
    }

    @Override
    public void flush() {
        if (futures.isEmpty()) return;

        try {
            CompletableFuture.allOf(futures.toArray(new CompletableFuture[0])).join();
            log.debug("[TraceFlow Async] Flushed {} events", eventCount.get());
            futures.clear();
            // All requests have settled; drop references to per-entity chains so the
            // map does not grow for the lifetime of the process.
            chains.clear();
            eventCount.set(0);
        } catch (Exception e) {
            if (silentErrors) {
                log.warn("[TraceFlow Async] Error during flush (silenced): {}", e.getMessage());
            } else {
                throw new TraceFlowException("Error during flush", e);
            }
        }
    }

    @Override
    public void shutdown() {
        log.debug("[TraceFlow Async] Shutting down async transport...");
        flush();
    }
}
